File: score/mw/com/impl/bindings/someip/sample_allocatee_ptr.cpp
/********************************************************************************
 * Copyright (c) 2026 Contributors to the Eclipse Foundation
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information regarding copyright ownership.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: Apache-2.0
 ********************************************************************************/
#include "score/mw/com/impl/bindings/someip/sample_allocatee_ptr.h"

#include <score/assert.hpp>

#include <utility>

namespace score::mw::com::impl::someip
{

SampleAllocateePtr::SampleAllocateePtr(std::nullptr_t /* ptr */) noexcept
    : managed_object_{nullptr}, event_slot_index_{kUninitialisedEventSlotIndex}, slot_allocation_control_{nullptr}
{
}

SampleAllocateePtr::SampleAllocateePtr(pointer ptr,
                                       SlotAllocationControl& slot_allocation_control,
                                       const SlotIndexType slot_index) noexcept
    : managed_object_{ptr}, event_slot_index_{slot_index}, slot_allocation_control_{&slot_allocation_control}
{
}

SampleAllocateePtr::SampleAllocateePtr(SampleAllocateePtr&& other) noexcept : SampleAllocateePtr()
{
    this->swap(other);
}

SampleAllocateePtr::~SampleAllocateePtr() noexcept
{
    internal_delete();
}

void SampleAllocateePtr::reset() noexcept
{
    internal_delete();
}

void SampleAllocateePtr::swap(SampleAllocateePtr& other) noexcept
{
    // Search for custom swap functions via ADL, and use std::swap if none are found.
    using std::swap;

    swap(this->managed_object_, other.managed_object_);
    swap(this->event_slot_index_, other.event_slot_index_);
    swap(this->slot_allocation_control_, other.slot_allocation_control_);
}

SampleAllocateePtr& SampleAllocateePtr::operator=(std::nullptr_t /* ptr */) & noexcept
{
    internal_delete();
    return *this;
}

SampleAllocateePtr& SampleAllocateePtr::operator=(SampleAllocateePtr&& other) & noexcept
{
    this->swap(other);
    return *this;
}

void SampleAllocateePtr::internal_delete() noexcept
{
    managed_object_ = nullptr;
    if (event_slot_index_ < kUninitialisedEventSlotIndex)
    {
        SCORE_LANGUAGE_FUTURECPP_PRECONDITION_PRD_MESSAGE(
            slot_allocation_control_ != nullptr,
            "The only time that slot_allocation_control_ is nullptr is if the SampleAllocateePtr is default "
            "initialised or "
            "initialised with a nullptr. In both these cases event_slot_index_ == kUninitialisedEventSlotIndex so we "
            "will never enter this branch.");
        slot_allocation_control_->DiscardSlot(event_slot_index_);
        event_slot_index_ = kUninitialisedEventSlotIndex;
    }
}

void swap(SampleAllocateePtr& lhs, SampleAllocateePtr& rhs) noexcept
{
    lhs.swap(rhs);
}

typename SampleAllocateePtr::pointer SampleAllocateePtrView::GetManagedObject() const noexcept
{
    return ptr_.managed_object_;
}

SlotAllocationControl& SampleAllocateePtrMutableView::GetSlotAllocationControl() const noexcept
{
    SCORE_LANGUAGE_FUTURECPP_PRECONDITION_PRD(ptr_.slot_allocation_control_ != nullptr);
    return *ptr_.slot_allocation_control_;
}

}  // namespace score::mw::com::impl::someip

File: score/mw/com/impl/plumbing/rust/sample_allocatee_ptr.rs
/********************************************************************************
 * Copyright (c) 2026 Contributors to the Eclipse Foundation
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information regarding copyright ownership.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: Apache-2.0
 ********************************************************************************/

///This crate gives same memory layout as SampleAllocateePtr in C++ implementation
///It can be used for FFI bindings where SampleAllocateePtr is needed
///It does not provide any functionality beyond memory layout compatibility
use core::fmt::Debug;
use std::mem::ManuallyDrop;

use common_rs::{
    BlankBinding,
    ConsumerEventDataControlLocalView,
    CxxOptional,
    CustomDeleter,
    ProviderEventDataControlLocalView,
    SlotIndexType,
};

#[repr(C)]
struct SampleAllocateeTracker {
    _data: [u8; 0],
}

#[repr(C)]
struct SampleAllocateeGuard {
    _tracker: *mut SampleAllocateeTracker,
}

#[repr(C)]
struct MockBinding<T> {
    _deleter: CustomDeleter,
    _ptr: *mut T,
}

#[repr(C)]
union MockBindingVariant<T> {
    _variant: ManuallyDrop<MockBinding<T>>,
}

#[repr(C)]
struct EventDataControlComposite {
    _asil_qm_control_local_: *mut ProviderEventDataControlLocalView,
    _asil_b_control_local_: *mut ProviderEventDataControlLocalView,
    _ignore_qm_control_: bool,
}

#[repr(C)]
struct LolaSampleAllocateePtrBinding<T> {
    _managed_object: *mut T,
    _event_slot_index: SlotIndexType,
    _event_data_control_ptr: *mut EventDataControlComposite,
    _consumer_data_control_local_: *mut ConsumerEventDataControlLocalView,
}

/// Mirror of `someip::SampleAllocateePtr`. Unlike the LoLa counterpart it is not templated on the sample type on
/// the C++ side, but the generic parameter is kept here so that all alternatives of the variant share one shape.
#[repr(C)]
struct SomeIpSampleAllocateePtrBinding<T> {
    _managed_object: *mut T,
    _event_slot_index: SlotIndexType,
    _owning_event: *mut core::ffi::c_void,
}

#[repr(C)]
union SomeIpSampleAllocateePtrVariant<T> {
    _variant: ManuallyDrop<SomeIpSampleAllocateePtrBinding<T>>,
    _mock_binding: ManuallyDrop<MockBindingVariant<T>>,
}

#[repr(C)]
union LolaSampleAllocateePtrVariant<T> {
    _variant: ManuallyDrop<LolaSampleAllocateePtrBinding<T>>,
    _someip_binding: ManuallyDrop<SomeIpSampleAllocateePtrVariant<T>>,
}

#[repr(C)]
union BlankVariant<T> {
    _variant: ManuallyDrop<BlankBinding>,
    _other: ManuallyDrop<LolaSampleAllocateePtrVariant<T>>,
}

#[repr(C)]
struct AllocationVariant<T> {
    _data: BlankVariant<T>,
    _index: u8,
}

#[repr(C, align(16))]
pub struct SampleAllocateePtr<T> {
    _internal: AllocationVariant<T>,
    _allocatee_guard: CxxOptional<SampleAllocateeGuard>,
}

// SAFETY: There is no connection of any data to a particular thread, also not on C++ side.
// Therefore, it is safe to send this struct to another thread.
unsafe impl<T> Send for SampleAllocateePtr<T> {}

impl<T> Debug for SampleAllocateePtr<T> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        // Must match the alternative order of
        // std::variant<score::cpp::blank, lola::SampleAllocateePtr, someip::SampleAllocateePtr,
        //              mock_binding::SampleAllocateePtr>
        // in score/mw/com/impl/plumbing/sample_allocatee_ptr.h.
        let state = match self._internal._index {
            0 => "Blank",
            1 => "LolaSampleAllocateePtr",
            2 => "SomeIpSampleAllocateePtr",
            3 => "UniquePtr",
            _ => "Unknown",
        };

        f.debug_struct("SampleAllocateePtr")
            .field("state", &state)
            .finish()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use test_helper_size_ffi_rs::SampleAllocateePtrLola;
    use test_utils_rs::*;

    #[test]
    fn test_sample_allocatee_ptr_variant_int32_size() {
        let cpp_size = SampleAllocateePtrLola::get_variant_int32();
        verify_size_and_align!(SampleAllocateePtr<i32>, cpp_size, "SampleAllocateePtr<i32>");
    }

    #[test]
    fn test_sample_allocatee_ptr_variant_u8_size() {
        let cpp_size = SampleAllocateePtrLola::get_variant_unsigned_char();
        verify_size_and_align!(SampleAllocateePtr<u8>, cpp_size, "SampleAllocateePtr<u8>");
    }

    #[test]
    fn test_sample_allocatee_ptr_variant_u64_size() {
        let cpp_size = SampleAllocateePtrLola::get_variant_unsigned_long_long();
        verify_size_and_align!(SampleAllocateePtr<u64>, cpp_size, "SampleAllocateePtr<u64>");
    }

    #[test]
    fn test_sample_allocatee_ptr_variant_user_defined_type_size() {
        let cpp_size = SampleAllocateePtrLola::get_variant_user_defined_type();
        verify_size_and_align!(
            SampleAllocateePtr<UserType>,
            cpp_size,
            "SampleAllocateePtr<UserType>"
        );
    }

    #[test]
    fn test_event_data_control_composite_size() {
        let cpp_size = SampleAllocateePtrLola::get_event_data_control_composite_size();
        verify_size_and_align!(
            EventDataControlComposite,
            cpp_size,
            "EventDataControlComposite"
        );
    }

    #[test]
    fn test_sample_allocatee_ptr_size() {
        let cpp_size = SampleAllocateePtrLola::get_sample_allocatee_ptr_size();
        verify_size_and_align!(
            LolaSampleAllocateePtrBinding<i32>,
            cpp_size,
            "LolaSampleAllocateePtrBinding<i32>"
        );
    }

    #[test]
    #[should_panic(expected = "size mismatch")]
    fn test_negative_allocatee_ptr_size_mismatch() {
        let cpp_size = SampleAllocateePtrLola::get_variant_int32();
        let incorrect = cpp_size.size + 1;
        assert_eq!(
            incorrect, cpp_size.size,
            "SampleAllocateePtr size mismatch!"
        );
    }

    #[test]
    #[should_panic(expected = "align mismatch")]
    fn test_negative_allocatee_ptr_align_mismatch() {
        let cpp_size = SampleAllocateePtrLola::get_variant_int32();
        let incorrect = cpp_size.align + 1;
        assert_eq!(
            incorrect, cpp_size.align,
            "SampleAllocateePtr align mismatch!"
        );
    }
}
